libraries/AP_ADSB/AP_ADSB_Sagetech.cpp

```diff
@@ -0,0 +1,684 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_ADSB_Sagetech.h"
+
+#if HAL_ADSB_ENABLED
+#include <GCS_MAVLink/GCS.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_RTC/AP_RTC.h>
+#include <stdio.h>
+#include <time.h>
+#include <string.h>
+#include <AP_HAL/utility/sparse-endian.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define SAGETECH_SCALER_LATLNG              (1.0f/2.145767E-5f)     //   180/(2^23)
+#define SAGETECH_SCALER_KNOTS_TO_CM         ((KNOTS_TO_M_PER_SEC/0.125f) * 100.0f)
+#define SAGETECH_SCALER_ALTITUDE            (1.0f/0.015625f)
+#define SAGETECH_SCALER_HEADING_CM          ((360.0f/256.0f) * 100.0f)
+
+#define SAGETECH_VALIDFLAG_LATLNG           (1<<0)
+#define SAGETECH_VALIDFLAG_ALTITUDE         (1<<1)
+#define SAGETECH_VALIDFLAG_VELOCITY         (1<<2)
+#define SAGETECH_VALIDFLAG_GND_SPEED        (1<<3)
+#define SAGETECH_VALIDFLAG_HEADING          (1<<4)
+#define SAGETECH_VALIDFLAG_V_RATE_GEO       (1<<5)
+#define SAGETECH_VALIDFLAG_V_RATE_BARO      (1<<6)
+#define SAGETECH_VALIDFLAG_EST_LATLNG       (1<<7)
+#define SAGETECH_VALIDFLAG_EST_VELOCITY     (1<<8)
+
+#define SAGETECH_ENFORCE_ACKS               0
+#define SAGETECH_DEBUG_ACK_TIMEOUTS         0
+#define SAGETECH_DEBUG_TX_ID_ONLY           0
+#define SAGETECH_DEBUG_TX_ID_PAYLOAD        0
+#define SAGETECH_DEBUG_TX_ALL_RAW           0
+#define SAGETECH_DEBUG_RX                   0
+#define SAGETECH_DEBUG_RX_ACK               0
+
+// detect if an port is configured as Sagetech
+bool AP_ADSB_Sagetech::detect()
+{
+    return (AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_Sagetech, 0) != nullptr);
+}
+
+bool AP_ADSB_Sagetech::init()
+{
+    if (_port != nullptr) {
+        return true;
+    }
+
+    _port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_Sagetech, 0);
+    if (_port == nullptr) {
+        return false;
+    }
+
+    baudrate = AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_Sagetech, 0);
+    _port->begin(baudrate);
+
+    // no sagtech hardware have flow control pins exposed
+    _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+
+    return true;
+}
+
+void AP_ADSB_Sagetech::update()
+{
+    if (_port == nullptr) {
+        return;
+    }
+
+    const uint32_t now_ms = AP_HAL::millis();
+
+    // -----------------------------
+    // read any available data on serial port
+    // -----------------------------
+    int32_t nbytes = _port->available();
+    while (nbytes-- > 0) {
+        const uint8_t data = (uint8_t)_port->read();
+
+        switch (protocol) {
+        default:

```

Comment:
it will be used in the multi-protocol parsing which I have now removed so this no longer applies. Since it takes 3 months to get PRs through I'm doing them one at a time to make them "faster"